examples/amm/src/contract.rs

```diff
@@ -0,0 +1,342 @@
+            AmmError::AmmChainOnly
+        );
+        match message {
+            Message::ExecuteOperation { operation } => {
+                self.execute_order_local(operation).await?;
+            }
+        }
+        Ok(ExecutionResult::default())
+    }
+
+    async fn handle_application_call(
+        &mut self,
+        context: &CalleeContext,
+        argument: Self::ApplicationCall,
+        _forwarded_sessions: Vec<SessionId>,
+    ) -> Result<ApplicationCallResult<Self::Message, Self::Response, Self::SessionState>, AmmError>
+    {
+        let mut result = ApplicationCallResult::default();
+        match argument {
+            ApplicationCall::ExecuteOperation { operation } => {
+                if context.chain_id == system_api::current_application_id().creation.chain_id {
+                    self.execute_order_local(operation).await?;
+                } else {
+                    self.execute_order_remote(&mut result.execution_result, operation)
+                        .await?;
+                }
+            }
+        }
+
+        Ok(result)
+    }
+
+    async fn handle_session_call(
+        &mut self,
+        _context: &CalleeContext,
+        _session: (),
+        _argument: (),
+        _forwarded_sessions: Vec<SessionId>,
+    ) -> Result<SessionCallResult<Self::Message, Self::Response, Self::SessionState>, AmmError>
+    {
+        Err(AmmError::SessionsNotSupported)
+    }
+}
+
+impl Amm {
+    async fn execute_order_local(&mut self, operation: OperationType) -> Result<(), AmmError> {
+        match operation {
+            OperationType::Swap {
+                owner,
+                input_token_idx,
+                input_amount,
+                output_token_idx,
+            } => {
+                if input_amount == Amount::ZERO {
+                    return Err(AmmError::NoZeroAmounts);
+                }
+
+                if input_token_idx > 1 || output_token_idx > 1 {
+                    return Err(AmmError::InvalidTokenIdx);
+                }
+
+                if input_token_idx == output_token_idx {
+                    return Err(AmmError::EqualTokensError);
+                }
+
+                let input_pool_balance = self.get_pool_balance(input_token_idx).await?;
+                let output_pool_balance = self.get_pool_balance(output_token_idx).await?;
+
+                let output_amount = amm::calculate_output_amount(
+                    input_amount,
+                    input_pool_balance,
+                    output_pool_balance,
+                )?;
+
+                self.receive_from_account(&owner, input_token_idx, input_amount)
+                    .await?;
+                self.send_to(&owner, output_token_idx, output_amount).await
+            }
+            OperationType::AddLiquidity {
+                owner,
+                token0_amount,
+                token1_amount,
+            } => {
```

Comment:
I'm now wondering whether in practice we can really expect the owner to use the right ratio here: If there are lots of users, the ratio will change in every swap.

Maybe the two amounts specified by the user should be taken as the maximal amounts they want to add. Then we compute the correct ratio and the tokens we actually add are such that they either match `token0_amount` and are less or equal `token1_amount`, or they match `token1_amount` and are less or equal `token0_amount`.

Then we could use the same principle in `RemoveLiquidity`.

Reply:
Good point, will adjust